src/python/pants/core/util_rules/external_tool.py

```diff
@@ -0,0 +1,221 @@
+            f'A "platform constraint" is any of {[c.value for c in PlatformConstraint]}, and '
+            "is the platform to fetch the tool for. A platform-independent tool should "
+            f"use {PlatformConstraint.none.value}, while a platform-dependent tool should specify "
+            'all environments it needs to be used for. The "fingerprint" and "size_bytes" '
+            "arguments are the result printed when running `sha256sum` and `wc -c` on "
+            "the downloaded file, respectively.",
+        )
+
+    @classmethod
+    @abstractmethod
+    def generate_url(cls, plat: Platform, version: str) -> str:
+        """Returns the URL for the given version of the tool, runnable on the given os+arch.
+
+        os and arch default to those of the current system.
+
+        Implementations should raise ExternalToolError if they cannot resolve the arguments
+        to a URL.  The raised exception need not have a message - a sensible one will be generated.
+        """
+
+    @classmethod
+    def generate_exe(cls, plat: Platform, version: str) -> str:
+        """Returns the archive path to the given version of the tool.
+
+        If the tool is downloaded directly, not in an archive, this can be left unimplemented.
+        """
+        return ""
+
+    @classmethod
+    def generate_request(
+        cls, plat: Platform, version: str, known_versions: List[str]
+    ) -> ExternalToolRequest:
+        for known_version in known_versions:
+            try:
+                ver, plat_val, sha256, length = (x.strip() for x in known_version.split("|"))
+            except ValueError:
+                raise ExternalToolError(
+                    f"Bad value for --known-versions (see ./pants "
+                    f"help-advanced {cls.options_scope}): {known_version}"
+                )
+            if plat_val == plat.value and ver == version:
+                digest = Digest(fingerprint=sha256, serialized_bytes_length=int(length))
+                try:
+                    url = cls.generate_url(plat, version)
+                    exe = cls.generate_exe(plat, version) or url.rsplit("/", 1)[-1]
+                except ExternalToolError as e:
+                    raise ExternalToolError(
+                        f"Couldn't find {cls.name} version {version} on {plat.value}"
+                    ) from e
+                return ExternalToolRequest(UrlToFetch(url=url, digest=digest), exe)
+        raise UnknownVersion(
+            f"No known version of {cls.name} {version} for {plat.value} found in {known_versions}"
+        )
+
+    def get_request(self, plat: Platform) -> ExternalToolRequest:
+        """Generate a request for this tool."""
+        return self.generate_request(
+            plat, self.get_options().version, self.get_options().known_versions,
+        )
+
+
+@rule
+async def download_external_tool(request: ExternalToolRequest) -> DownloadedExternalTool:
```

Comment:
So, it's currently the case that the Platform is available as a singleton, here: https://github.com/pantsbuild/pants/blob/aa2ed8b9b200122ed47d222073125d315de34bf3/src/python/pants/engine/platform.py#L32-L36 ... that means that you don't need to take the Platform as an argument as part of the Request, and can instead request the platform singleton.

So your rule signature could probably just be:
```
@rule
async def download_external_tool(request: ExternalToolRequest, platform: Platform) -> DownloadedExternalTool:
```

The plan shortly (via https://docs.google.com/document/d/1hS0n6vp-hBxy4Xo-q9QA_elofzQtOubjfN4nSeIgp90/edit#heading=h.jj08j7tn7x2d , which I have some initial patches for) is for the Platform to switch from being a singleton to being a Param. That won't change anything from the `@rule`'s perspective, but it will actually allow differentiating requests from one another.

Reply:
Gotcha. Will make this change.

Reply:
Hmm, so we need to know the Platform when we create the request, because it can affect the URL, so this would require more restructuring than I think we want right now?

Reply:
Not a blocker. Thanks!